### PR TITLE
exp/agents: flesh out agents implementation

### PR DESCRIPTION
**Pull request by Abraxas-365**
#37 Hi this is my approach on agent and Zero shot agent, Already have the _test but i need my tools PR to be merged to work

### CODE REVIEW

File: exp/agents/agentExecutor/agentExecutor.go [hunk 1]
```diff
@@ -0,0 +1,7 @@
+package agentExecutor
```
Comment (tmc) [inline exp/agents/agentExecutor/agentExecutor.go:1]: go files and package names should only consist of lowercase characters, see https://github.com/uber-go/guide/blob/master/style.md#package-names

File: exp/agents/agentExecutor/agentExecutor.go [hunk 2]
```diff
@@ -0,0 +1,7 @@
...
+
+import "github.com/tmc/langchaingo/schema"
+
+type AgentExecutor interface {
```
Comment (tmc) [inline exp/agents/agentExecutor/agentExecutor.go:5]: new exported types should have comments.

File: exp/agents/agents.go [hunk 3]
```diff
@@ -0,0 +1,51 @@
+// Package agents defines the types for langchaingo Agetns.
```
Comment (tmc) [inline exp/agents/agents.go:1]: please place package-level comments into a doc.go file

File: exp/agents/agents.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+package agents
+
+import (
+	"errors"
+
+	agentsExecutor "github.com/tmc/langchaingo/exp/agents/agentExecutor"
```
Comment (tmc) [inline exp/agents/agents.go:7]: why alias this package?

File: exp/agents/agents.go [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+	"github.com/tmc/langchaingo/llms"
+)
+
+type AgentType string
+
+const (
+	ZeroShotReactDescription AgentType = "zeroShotReactDescription"
+)
+
+type AgentOptions map[string]any
```
Comment (tmc) [inline exp/agents/agents.go:19]: why this type?

File: exp/agents/agents.go [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+type Options func(p *AgentOptions)
```
Comment (tmc) [inline exp/agents/agents.go:20]: let's call this "AgentOption"

File: exp/agents/agents.go [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+import (
+	"errors"
+
+	agentsExecutor "github.com/tmc/langchaingo/exp/agents/agentExecutor"
+	"github.com/tmc/langchaingo/exp/agents/mrkl"
+	"github.com/tmc/langchaingo/exp/tools"
+	"github.com/tmc/langchaingo/llms"
+)
+
+type AgentType string
```
Comment (tmc) [inline exp/agents/agents.go:13]: exported symbols should have comments.

File: exp/agents/agents.go [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+
+const (
+	ZeroShotReactDescription AgentType = "zeroShotReactDescription"
+)
+
+type AgentOptions map[string]any
+type Options func(p *AgentOptions)
+
+func defaultOptions() AgentOptions {
```
Comment (tmc) [inline exp/agents/agents.go:22]: is this necessary?

File: exp/agents/agents.go [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+	}
+}
+
+func WithVerbosity() Options {
+	return func(p *AgentOptions) {
+		(*p)["verbose"] = true
+	}
+}
+
+func InitializeAgent(
```
Comment (tmc) [inline exp/agents/agents.go:34]: why not "NewAgent" ?

File: exp/agents/mrkl/mrkl.go [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+	"strings"
+
+	"github.com/tmc/langchaingo/exp/agents/agentExecutor"
+	"github.com/tmc/langchaingo/exp/chains"
+	"github.com/tmc/langchaingo/exp/tools"
+	"github.com/tmc/langchaingo/llms"
+	"github.com/tmc/langchaingo/schema"
+)
+
+type OneShotZeroAgent struct {
```
Comment (tmc) [inline exp/agents/mrkl/mrkl.go:14]: exported symbols should have comments.

File: exp/agents/mrkl/mrkl.go [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+	llm        llms.LLM
+	query      string
+	chain      chains.Chain
+	tools      []tools.Tool
+	verbose    bool
+	maxRetries int
+}
+
+type OneShotZeroAgentOptions map[string]any
```
Comment (tmc) [inline exp/agents/mrkl/mrkl.go:23]: exported symbols should have comments.

File: exp/agents/mrkl/mrkl.go [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+
+var _ agentExecutor.AgentExecutor = (*OneShotZeroAgent)(nil)
+
+const FinalAnswerAction = "Final Answer:"
```
Comment (tmc) [inline exp/agents/mrkl/mrkl.go:27]: exported symbols should have comments.

File: exp/agents/mrkl/mrkl.go [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+	if _, ok := opts["verbose"].(bool); !ok {
+		opts["verbose"] = false
+	}
+	if _, ok := opts["maxRetries"].(int); !ok {
+		opts["maxRetries"] = 3
+	}
+	return opts
+}
+
+//Aqui se crea el chain
```
Comment (tmc) [inline exp/agents/mrkl/mrkl.go:39]: please put comments into the form of " NewShotAgent ..." and have the comments be in english.

File: exp/agents/mrkl/mrkl.go [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+		output, err = a.nextStep(*action)
+		if err != nil {
+			return nil, err
+		}
+
+		attempts++
+	}
+	return nil, fmt.Errorf("Agent did not finish after %d attempts", attempts)
+}
+func (a *OneShotZeroAgent) nextStep(action schema.AgentAction) (string, error) {
```
Comment (tmc) [inline exp/agents/mrkl/mrkl.go:83]: missing a newline before this

File: exp/agents/mrkl/prompt.go [hunk 15]
```diff
@@ -0,0 +1,101 @@
+package mrkl
```
Comment (tmc) [inline exp/agents/mrkl/prompt.go:1]: new packages should have doc.go files with a package-evel comment

File: exp/agents/mrkl/prompt.go [hunk 16]
```diff
@@ -0,0 +1,101 @@
...
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/tmc/langchaingo/exp/prompts"
+	"github.com/tmc/langchaingo/exp/tools"
+)
+
+const PREFIX = `Answer the following questions as best you can. You have access to the following Tools:`
```
Comment (tmc) [inline exp/agents/mrkl/prompt.go:11]: prefer a `const (` expression,

Do not do full-caps exports of constants, instead use TitleCase (and comment the export) or lowerCamel.

File: exp/agents/mrkl/prompt.go [hunk 17]
```diff
@@ -0,0 +1,101 @@
...
+Observation: the result of the action
+... (this Thought/Action/Action Input/Observation can repeat N times)
+Thought: I now know the final answer
+Final Answer: the final answer to the original input question`
+const SUFFIX = `Begin!
+
+Question: {input}
+Thought:{agent_scratchpad}`
+
+type CreatePromptOptions struct {
```
Comment (tmc) [inline exp/agents/mrkl/prompt.go:27]: exported symbols should have comments.

File: exp/agents/mrkl/prompt.go [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+	Prefix             string
+	Suffix             string
+	FormatInstructions string
+	InputVariables     []string
+}
+
+type PromptTemplateOption func(p *CreatePromptOptions)
```
Comment (tmc) [inline exp/agents/mrkl/prompt.go:34]: exported symbols should have comments.

File: exp/agents/mrkl/prompt.go [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+
+func WithPrefix(prefix string) PromptTemplateOption {
```
Comment (tmc) [inline exp/agents/mrkl/prompt.go:36]: exported symbols should have comments.

File: exp/tools/serpapi/internal/serpapiClient.go [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+)
+
+var ErrMissingToken = errors.New("missing the OpenAI API key, set it in the OPENAI_API_KEY environment variable")
```
Comment (tmc) [inline exp/tools/serpapi/internal/serpapiClient.go:15]: prefer a `var (` expression

File: exp/tools/serpapi/internal/serpapiClient.go [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+var NoGoodResult = "No good search result found"
+
+type SerpapiClient struct {
```
Comment (tmc) [inline exp/tools/serpapi/internal/serpapiClient.go:18]: exported symbols should have comments.

File: exp/tools/serpapi/internal/serpapiClient.go [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+	apiKey  string
+	baseURL string
+}
+
+func New() (*SerpapiClient, error) {
```
Comment (tmc) [inline exp/tools/serpapi/internal/serpapiClient.go:23]: exported symbols should have comments.

File: exp/tools/serpapi/internal/serpapiClient.go [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+	}
+
+	return &SerpapiClient{
+			apiKey:  apiKey,
+			baseURL: "https://serpapi.com/search",
+		},
+		nil
+}
+
+func (s *SerpapiClient) Search(query string) (string, error) {
```
Comment (tmc) [inline exp/tools/serpapi/internal/serpapiClient.go:36]: exported symbols should have comments.

File: exp/tools/tool.go [hunk 24]
```diff
@@ -0,0 +1,17 @@
...
+package tools
+
+// Tool is a tool for the llm agent to interact with diferent application.
+type Tool struct {
+	Name        string
+	Description string
+	Run         func(string) string
+}
+
+func NewTool(name, description string, run func(string) string) *Tool {
```
Comment (tmc) [inline exp/tools/tool.go:11]: exported symbols should have comments.

File: exp/tools/tool.go [hunk 25]
```diff
@@ -0,0 +1,17 @@
+// Package tools defines the types for tools to be used by the llms agents.
```
Comment (tmc) [inline exp/tools/tool.go:1]: place package-level documentation into a doc.go file

File: llms/openai/openaillm.go [hunk 26]
```diff
@@ -34,7 +34,8 @@ func (o *LLM) Call(prompt string, stopWords []string) (string, error) {
 
 func (o *LLM) Generate(prompts []string, stopWords []string) ([]*llms.Generation, error) {
 	result, err := o.client.CreateCompletion(context.TODO(), &openaiclient.CompletionRequest{
-		Prompt: prompts[0],
+		Prompt:    prompts[0],
+		StopWords: stopWords,
```
Comment (tmc) [inline llms/openai/openaillm.go:38]: good catch.

**Review by tmc — COMMENTED** [review]
Overall pretty good but left a number of comments.

**Comment by tmc** [comment]
This looks pretty good but we should be wary about adding a lot of new content under exp -> this directory is meant as a staging location for lower quality code and I'd prefer to get contributions directly into the top level packages.

File: exp/agents/agents.go (hunk 4 above)
Comment (Abraxas-365) [reply exp/agents/agents.go:7]: My bad, sorry, will fix it

File: exp/agents/agents.go (hunk 5 above)
Comment (Abraxas-365) [reply exp/agents/agents.go:19]: I tried to made it as similar as chains ```go

**Comment by Abraxas-365** [comment]
Thanks for the commenrts , will fix it , just realize _exp_ means experimental hahahaha

File: exp/agents/mrkl/mrkl.go (hunk 13 above)
Comment (Abraxas-365) [reply exp/agents/mrkl/mrkl.go:39]: Sorryy hahhaa , forgot to delete that ahahah

File: exp/agents/agents.go (hunk 3 above)
Comment (ChristianSch) [reply exp/agents/agents.go:1]: also there's a typo in "Agetns"

File: exp/tools/serpapi/internal/serpapiClient.go (hunk 20 above)
Comment (ChristianSch) [reply exp/tools/serpapi/internal/serpapiClient.go:15]: there's a copy and paste error, it's not the openAPI key that's missing ;)

Comment (Abraxas-365) [reply exp/tools/serpapi/internal/serpapiClient.go:15]: Sorry i fixed it in the Tools PR, forgot to change it here #19

**Comment by Abraxas-365** [comment]
Hi pls close my Tools PR and just accept this one , this contains Agent and Tools changes I had to do a merge, to finish the develop of the package

File: exp/agent/executor/agentExecutor.go [hunk 27]
```diff
@@ -0,0 +1,11 @@
+package executor
```
Comment (tmc) [inline exp/agent/executor/agentExecutor.go:1]: go file names should be snake_case. include a doc.go in each package.

**Review by tmc — COMMENTED** [review]
Almost there, please address these last comments.

File: exp/agent/executor/agentExecutor.go (hunk 27 above)
Comment (Abraxas-365) [reply exp/agent/executor/agentExecutor.go]: hahahaha I just made the changes

**Review by tmc — APPROVED** [review]
LGTM

**Pull request merged**